m3/src/test/java/com/uber/m3/tally/m3/MockM3Server.java

```
@@ -0,0 +1,88 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.uber.m3.tally.m3;
+
+import com.uber.m3.tally.m3.thrift.TUdpServer;
+import com.uber.m3.thrift.generated.M3;
+import org.apache.thrift.TException;
+import org.apache.thrift.TProcessor;
+import org.apache.thrift.protocol.TCompactProtocol;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.transport.TTransport;
+import org.apache.thrift.transport.TTransportException;
+
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.util.concurrent.Phaser;
+
+public class MockM3Server {
+    private TProcessor processor;
+    private TTransport transport;
+    private MockM3Service service;
+
+    public MockM3Server(
+        Phaser phaser,
+        boolean countBatches,
+        SocketAddress address
+    ) {
+        service = new MockM3Service(phaser, countBatches);
+
+        processor = new M3.Processor<>(service);
+
+        try {
+            transport = new TUdpServer(address);
+        } catch (SocketException e) {
+            throw new RuntimeException("Unable to open socket", e);
+        }
+    }
+
+    public void serve() {
+        try {
+            transport.open();
+        } catch (TTransportException e) {
+            throw new RuntimeException("Failed to open socket", e);
+        }
+
+        TProtocol protocol = new TCompactProtocol.Factory().getProtocol(transport);
+
+        while (transport.isOpen()) {
+            try {
+                processor.process(protocol, protocol);
+            } catch (TException e) {
+                if (transport.isOpen()) {
+                    // If we reach here and the transport is still open, then
+                    // something bad probably happened.
+                    e.printStackTrace();
```

do you want to throw a runtime exception here?